File: stdlib/Collections.java
package stdlib;

import java.util.*;
import java.util.function.*;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.core.type.TypeReference;

public class Collections {
    private static final ObjectMapper mapper = new ObjectMapper();

    public static class Collection<T> {
        protected String type;
        protected Object data;

        protected Collection(String type, Object data) {
            this.type = type;
            this.data = data;
        }

        public String toJson() throws Exception {
            Map<String, Object> json = new HashMap<>();
            json.put("type", type);
            json.put("data", data);
            return mapper.writeValueAsString(json);
        }
    }

    public static class CPJList<T> extends Collection<T> {
        private List<T> items;

        public CPJList() {
            super("list", new ArrayList<T>());
            this.items = new ArrayList<>();
        }

        public CPJList(List<T> items) {
            super("list", items);
            this.items = items;
        }

        public static <T> CPJList<T> fromJson(String jsonStr) throws Exception {
            Map<String, Object> json = mapper.readValue(jsonStr, new TypeReference<Map<String, Object>>() {
            });
            if (!"list".equals(json.get("type"))) {
                throw new IllegalArgumentException("Invalid collection type");
            }
            List<T> items = mapper.convertValue(json.get("data"), new TypeReference<List<T>>() {
            });
            return new CPJList<>(items);
        }

        public void append(T item) {
            items.add(item);
        }

        public <R> CPJList<R> map(Function<T, R> func) {
            List<R> result = new ArrayList<>();
            for (T item : items) {
                result.add(func.apply(item));
            }
            return new CPJList<>(result);
        }

        public CPJList<T> filter(Predicate<T> pred) {
            List<T> result = new ArrayList<>();
            for (T item : items) {
                if (pred.test(item)) {
                    result.add(item);
                }
            }
            return new CPJList<>(result);
        }
    }

    public static class CPJDict<K, V> extends Collection<Map.Entry<K, V>> {
        private Map<K, V> items;

        public CPJDict() {
            super("dict", new HashMap<K, V>());
            this.items = new HashMap<>();
        }

        public static <K, V> CPJDict<K, V> fromJson(String jsonStr) throws Exception {
            Map<String, Object> json = mapper.readValue(jsonStr, new TypeReference<Map<String, Object>>() {
            });
            if (!"dict".equals(json.get("type"))) {
                throw new IllegalArgumentException("Invalid collection type");
            }
            Map<K, V> items = mapper.convertValue(json.get("data"), new TypeReference<Map<K, V>>() {
            });
            CPJDict<K, V> result = new CPJDict<>();
            items.forEach(result::set);
            return result;
        }

        public void set(K key, V value) {
            items.put(key, value);
        }

        public V get(K key, V defaultValue) {
            return items.getOrDefault(key, defaultValue);
        }
    }

    public static class CPJSet<T> extends Collection<T> {
        private Set<T> items;

        public CPJSet() {
            super("set", new HashSet<T>());
            this.items = new HashSet<>();
        }

        public CPJSet(Set<T> items) {
            super("set", items);
            this.items = items;
        }

        public static <T> CPJSet<T> fromJson(String jsonStr) throws Exception {
            Map<String, Object> json = mapper.readValue(jsonStr, new TypeReference<Map<String, Object>>() {
            });
            if (!"set".equals(json.get("type"))) {
                throw new IllegalArgumentException("Invalid collection type");
            }
            List<T> list = mapper.convertValue(json.get("data"), new TypeReference<List<T>>() {
            });
            return new CPJSet<>(new HashSet<>(list));
        }

        public void add(T item) {
            items.add(item);
        }

        public void remove(T item) {
            items.remove(item);
        }

        public CPJSet<T> union(CPJSet<T> other) {
            Set<T> result = new HashSet<>(items);
            result.addAll(other.items);
            return new CPJSet<>(result);
        }
    }

    // Example usage
    public static void main(String[] args) throws Exception {
        // Create a list
        CPJList<Integer> numbers = new CPJList<>();
        numbers.append(1);
        numbers.append(2);
        numbers.append(3);

        // Map and filter operations
        CPJList<Integer> doubled = numbers.map(x -> x * 2);
        CPJList<Integer> evens = doubled.filter(x -> x % 2 == 0);

        // Convert to JSON
        String json = evens.toJson();
        System.out.println("JSON: " + json);

        // Create from JSON
        CPJList<Integer> received = CPJList.fromJson(json);

        // Dictionary example
        CPJDict<String, int[]> points = new CPJDict<>();
        points.set("origin", new int[] { 0, 0 });
        points.set("end", new int[] { 100, 100 });

        // Set operations
        CPJSet<Integer> set1 = new CPJSet<>();
        set1.add(1);
        set1.add(2);
        set1.add(3);

        CPJSet<Integer> set2 = new CPJSet<>();
        set2.add(3);
        set2.add(4);
        set2.add(5);

        CPJSet<Integer> union = set1.union(set2);
    }
}

File: stdlib/collections.py
"""
CPJ Standard Library - Collections Module
Implements core collection types that work seamlessly across C++, Python, and Java
"""

from typing import TypeVar, Generic, List, Dict, Set
import json
from dataclasses import dataclass

T = TypeVar('T')
K = TypeVar('K')
V = TypeVar('V')

@dataclass
class CPJCollection(Generic[T]):
    """Base class for all CPJ collections with cross-language support"""
    _data: List[T]
    _type: str

    def to_json(self) -> str:
        """Convert to JSON for cross-language transfer"""
        return json.dumps({
            "type": self._type,
            "data": self._data
        })

    @classmethod
    def from_json(cls, json_str: str) -> 'CPJCollection[T]':
        """Create collection from JSON (used by C++ and Java runtimes)"""
        data = json.loads(json_str)
        return cls(data["data"], data["type"])

class CPJList(CPJCollection[T]):
    """List implementation that works across all three languages"""
    def __init__(self, items: List[T] = None):
        super().__init__(items or [], "list")

    def append(self, item: T) -> None:
        self._data.append(item)

    def extend(self, other: 'CPJList[T]') -> None:
        self._data.extend(other._data)

    def map(self, fn) -> 'CPJList[T]':
        """Higher-order function support"""
        return CPJList([fn(x) for x in self._data])

    def filter(self, pred) -> 'CPJList[T]':
        return CPJList([x for x in self._data if pred(x)])

class CPJDict(CPJCollection[Dict[K, V]]):
    """Dictionary implementation with cross-language support"""
    def __init__(self, items: Dict[K, V] = None):
        super().__init__(items or {}, "dict")

    def set(self, key: K, value: V) -> None:
        self._data[key] = value

    def get(self, key: K, default: V = None) -> V:
        return self._data.get(key, default)

class CPJSet(CPJCollection[T]):
    """Set implementation that works in all three languages"""
    def __init__(self, items: Set[T] = None):
        super().__init__(set(items or []), "set")

    def add(self, item: T) -> None:
        self._data.add(item)

    def remove(self, item: T) -> None:
        self._data.remove(item)

    def union(self, other: 'CPJSet[T]') -> 'CPJSet[T]':
        return CPJSet(self._data.union(other._data))

# Example usage in Python:
def example():
    # Create a list
    numbers = CPJList([1, 2, 3, 4, 5])
    
    # Map and filter operations
    doubled = numbers.map(lambda x: x * 2)
    evens = doubled.filter(lambda x: x % 2 == 0)
    
    # Convert to JSON for C++ or Java
    json_data = evens.to_json()
    
    # Create from JSON (simulating C++ or Java sending data)
    received = CPJList[int].from_json(json_data)
    
    # Dictionary example
    points = CPJDict[str, tuple]()
    points.set("origin", (0, 0))
    points.set("end", (100, 100))
    
    # Set operations
    set1 = CPJSet([1, 2, 3])
    set2 = CPJSet([3, 4, 5])
    union = set1.union(set2)
    
    return {
        "list": received._data,
        "dict": points._data,
        "set": union._data
    }